modules/core/04-channel/v2/keeper/msg_server_test.go

```diff
@@ -433,3 +435,112 @@ func (suite *KeeperTestSuite) TestMsgAcknowledgement() {
 		})
 	}
 }
+
+func (suite *KeeperTestSuite) TestMsgTimeout() {
+	var (
+		path       *ibctesting.Path
+		msgTimeout *channeltypesv2.MsgTimeout
+		packet     channeltypesv2.Packet
+	)
+
+	testCases := []struct {
+		name     string
+		malleate func()
+		expError error
+	}{
+		{
+			name:     "success",
+			malleate: func() {},
+		},
+		{
+			name: "failure: no-op",
+			malleate: func() {
+				suite.chainA.App.GetIBCKeeper().ChannelKeeperV2.DeletePacketCommitment(suite.chainA.GetContext(), packet.DestinationChannel, packet.Sequence)
+
+				// Modify the callback to return a different error.
+				// This way, we can verify that the callback is not executed in a No-op case.
+				path.EndpointA.Chain.GetSimApp().MockModuleV2A.IBCApp.OnTimeoutPacket = func(context.Context, string, string, channeltypesv2.PacketData, sdk.AccAddress) error {
+					return errors.New("OnTimeoutPacket callback failed")
+				}
+			},
+			expError: channeltypesv1.ErrNoOpMsg,
+		},
+		{
+			name: "failure: invalid signer",
+			malleate: func() {
+				msgTimeout.Signer = ""
+			},
+			expError: errors.New("empty address string is not allowed"),
+		},
+		{
+			name: "failure: callback fails",
+			malleate: func() {
+				path.EndpointA.Chain.GetSimApp().MockModuleV2A.IBCApp.OnTimeoutPacket = func(context.Context, string, string, channeltypesv2.PacketData, sdk.AccAddress) error {
+					return errors.New("OnTimeoutPacket callback failed")
+				}
+			},
+			expError: errors.New("OnTimeoutPacket callback failed"),
+		},
+		{
+			name: "failure: channel not found",
+			malleate: func() {
+				// change the source id to a non-existent channel.
+				msgTimeout.Packet.SourceChannel = "not-existent-channel"
+			},
+			expError: channeltypesv2.ErrChannelNotFound,
+		},
+		{
+			name: "failure: invalid commitment",
+			malleate: func() {
+				suite.chainA.App.GetIBCKeeper().ChannelKeeperV2.SetPacketCommitment(suite.chainA.GetContext(), packet.SourceChannel, packet.Sequence, []byte("foo"))
+			},
+			expError: channeltypesv2.ErrInvalidPacket,
+		},
+		{
+			name: "failure: failed membership verification",
+			malleate: func() {
+				msgTimeout.ProofHeight = clienttypes.ZeroHeight()
+			},
+			expError: clienttypes.ErrConsensusStateNotFound,
+		},
+	}
+	for _, tc := range testCases {
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+
+			path = ibctesting.NewPath(suite.chainA, suite.chainB)
+			path.SetupV2()
+
+			// Send packet from A to B
+			timeoutTimestamp := suite.chainA.GetTimeoutTimestamp()
+
+			mockData := mockv2.NewMockPacketData(mockv2.ModuleNameA, mockv2.ModuleNameB)
+			msgSendPacket := channeltypesv2.NewMsgSendPacket(path.EndpointA.ChannelID, timeoutTimestamp, suite.chainA.SenderAccount.GetAddress().String(), mockData)
```

Comment:
we should be able to use the new helper fn for this now (might need to merge feature branch first)